Get working with web-sys

tarkah opened the pull request:
Hey! I've been really wanting to switch over to using the newer web-sys branch for yew, but yew_router was holding me back. So I attempted a super rough attempt at adding a web_sys vs std_web feature here, which satisfies #215.

It seems to work! I've branched my personal project to use this here, [https://github.com/tarkah/karaoke-rs/pull/19](https://github.com/tarkah/karaoke-rs/pull/19), and using both `std_web` and `web_sys` features work. Using `web_sys` has actually solved some issues I was having w/ yew stdweb since I am using web-sys for some WebAudio and Canvas API stuff.

I didn't really know how to handle `RouteState`, but outside that I think it's _mostly_ correct. Hopefully is helpful and you can easily clean up anything I may have done wrong.

tarkah committed:
Get working with web-sys

tarkah commented:
Only thing I am seeing is a warning of
```
Ignoring ‘preventDefault()’ call on event of type ‘click’ from a listener registered as ‘passive’.
```

Which must relate to:

https://github.com/tarkah/yew_router/blob/2e93940c9164f1efe1d4ddeebb88784abd7918fc/src/components/router_link.rs#L71-L73

Gloo events has a warning about this, not sure if it relates at all: https://docs.rs/gloo-events/0.1.1/gloo_events/struct.EventListener.html#passive. No idea if this is a big issue or not

hgzimmerman commented:
Thanks a ton for taking the initiative on this!
RouteState is a thing I need to resolve in a general sense, so it isn't really an impediment to getting this PR merged, even if what you have doesn't work.

I'm a little sick so I'll take a look at this if I have the energy after work.

tarkah commented:
No problem! I've had enough exposure to wasm-bindgen over the last couple weeks where I felt confident tackling this.

No rush, hope you feel better!

tarkah commented:
We need to disable `href` for `RouterAnchor` for web-sys until we can get the underlying `onclick` event listener to be created with options to enable `preventDefault()` in Chrome / Firefox. Without this option working, the page is getting reloaded.

I'll open up an issue on Yew to see if we can maybe get a new event listener type `onclick_nonpassive?` that has this feature enabled

Again, see:

https://docs.rs/gloo-events/0.1.1/gloo_events/struct.EventListener.html#passive

jstarry commented:
@tarkah you could also use refs and use web_sys directly to attach an event listener with your required listener options. Doesn't require a yew change I think.

tarkah committed:
Use custom event listener for web-sys

Needed to create event listener in non-passive state so
`prevent_default` works in Chrome / Firefox, so page doesn't get
reloaded when html link is clicked.

tarkah commented:
@jstarry Thanks for the recommendation on refs! I didn't know about NodeRef and that ended up working perfectly!

hgzimmerman commented:
I've done a quick skim over and as far as I can tell, it looks good.

I'd like to get the CI working again though.

It looks like the examples are currently failing with the error: 
```
'Yew requires selecting either the `web_sys` or `std_web` cargo feature'
```
for the Yew dependency.

Would you mind specifying the proper features in the examples? I think `stdweb` should be fine, as that won't require any additional effort on your part aside from specifying the features.

tarkah commented:
Awesome, thanks for reviewing. I'll update the examples and any other CI issues that come up.

tarkah committed:
update CI for web_sys / std_web features

tarkah commented:
Alright, this should be good to go. I've also updated an outdated dependency on the `minimal` example that was preventing it from building - `web_logger` from 0.1 to 0.2.
